### ISSUE TITLE
floccus becomes unusuable on Firefox and stays in syncing mode

### ISSUE THREAD

**Issue by Whissi** [floccusaddon/floccus#1242]
### Which version of floccus are you using?

4.17.0

### Sync method

Nextcloud Bookmarks

### Which browser are you using? In case you are using the phone App, specify the Android or iOS version and device please.

Mozilla Firefox 104.0 64-bit

### Which version of Nextcloud Bookmarks are you using? (if relevant)

v11.0.1

### Which version of Nextcloud? (if relevant)

v24.0.4

### Describe the Bug

For some months now I notice that the bookmark icon in Firefox's nav bar doesn't work anymore: When I click on the star, a menu like

![firefox-add-bookmark-dialog](https://user-images.githubusercontent.com/729436/187098074-a0d39851-cf6f-46a4-9e21-56608429577f.png)

should appear. However, nothing is happening.

When I disable Floccus add-on, bookmark button in nav bar is working as expected.

When button is not working, it will also disable "null" as tooltip when I hover the icon while normally it should display "Bookmark this page (CTRL+D)".

Pressing CTRL+D will bring up the menu.

### Expected Behavior

The star icon in Firefox's nav bar should work, i.e.

- Should be colored when you visit an already bookmarked page
- Should display a menu when you to add or edit bookmark

### To Reproduce

1. Start Firefox
2. Visit any bookmarked page, check if star icon in nav bar is just shown as contours of a star or if you see a colored star (the latter will indicate that you are viewing a bookmarked page).
3. Also, try to click the star icon in nav bar and see if you get the bookmark add/edit dialog.

### Debug log provided

- [ ] I have provided a debug log file

**Comment by Whissi** [comment]
I just noticed that Floccus seems to stay in "Syncing" status, see

![floccus-always-synchronizing](https://user-images.githubusercontent.com/729436/187098472-67269840-5caf-4c5e-b8c3-fb7c3500f21e.png)

However, last line of log is

> 2022-08-28T01:53:12.860Z Successfully ended sync process for account myaccount@example.org

Not sure if this has something to do with the problem.

**Comment by Whissi** [comment]
I clicked "Reset Cache" in options and both issues are gone for now, i.e. add-on shows "All good" and bookmark icon in nav bar is working again.

I leave this open since I had the same experience when disabling/re-enabling floccus add-on but error comes back after few hours/days.

**Comment by marcelklehr** [comment]
Did the error come back?

**Comment by Whissi** [comment]
Yes, was working for some days but I am experiencing the problem right now.

**Comment by Whissi** [comment]
Today I hit this again.

**Comment by Whissi** [comment]
Today it happened again.

**Comment by marcelklehr** [comment]
I can confirm this issue, but I doubt it's something that can be fixed in floccus.

**Comment by Whissi** [comment]
Do you have more information? I don't know when it started but there were versions without this problem.
At the moment this problem is making floccus unusable for me like I just noticed you cannot edit/delete any bookmarks in Firefox bookmark windows (CTRL+SHIFT+O) when floccus is causing Firefox to stays in "syncing" mode. I am also wondering about the impact of resetting cache every 1-2 days (i.e. is floccus for example not deleting bookmarks on server I changed/deleted locallay because UI says "the next synchronization run is guaranteed not to delete any data and merely merges server and local bookmarks together").

**Comment by marcelklehr** [comment]
> like I just noticed you cannot edit/delete any bookmarks in Firefox bookmark windows (CTRL+SHIFT+O) when floccus is causing Firefox to stays in "syncing" mode

I cannot reproduce this. What is "syncing" mode?

The only thing I can reproduce is that the bookmark button in the firefox URL bar doesn't correctly indicate whether the current site is bookmarked.

> I am also wondering about the impact of resetting cache every 1-2 days 

I am not sure why resetting the cache would fix this issue temporarily. I think it's likely a coincidence and it's actually an intermittent issue.

**Comment by Whissi** [comment]
With "syncing mode" I mean the UI status shown in the screenshot above.

**Comment by Whissi** [comment]
I updated issue to reflect the main issue.

When I initially reported this issue I thought that this was only an UI issue, i.e. only affects icon shown on nav bar and maybe bookmark menu. Nothing that critical. However, that's not the case. Add-on is completely not usable for me when this error occurs and floccus stays in "syncing" mode. I.e. not just UI is blocked, even new or changed bookmarks don't get synchronized anymore (shouldn't be surprising given that add-on stays in "syncing" status).

**Comment by marcelklehr** [comment]
So, this is about two issues as far as I can see, then:

1) Bookmark Button in Nav bar not working sometimes
2) Floccus gets stuck syncing (so similar to https://github.com/floccusaddon/floccus/issues/1252 perhaps)

Can you open a new issue for 2, so we can separate the two? The bookmark button not working is likely a firefox bug and nothing I can do about in floccus.

**Comment by Whissi** [comment]
I don't believe that these are two issues. I believe that the thing with the button in navbar or not working bookmark dialog is caused by floccus getting stuck in syncing mode which will block bookmark API from firefox affecting firefox own navbar and other features like the bookmark dialog which all need to query bookmark data.

**Comment by flexy2dd** [comment]
Hi all
I have the exact same issue (with 4.17.1) and I agree with @Whissi is not just a UI issue but it's sync loop
and is not possible to cancel the sync, if click on cancel button no effect

**Comment by Whissi** [comment]
I just noticed that this bug is causing data loss:

I disabled auto-sync for a while in hope that Firefox will keep working until I manually sync and have to clear cache in floccus afterwards to unbreak it. However, it looks like this is not enough.

Today for example I added a new bookmark. I noticed that UI was stuck (the way using the bookmark star). But I was able to create the bookmark via context menu. It was visible in my bookmark toolbar.

I restarted the complete browser and after restart I noticed that the new bookmark was gone.

I now moved one bookmark from folder A to folder B. Tried to sync in floccus but "Sync now" button didn't work (nothing happened when I clicked the button). So I cleared cache in floccus, restarted Firefox and all my changes were lost, i.e. the bookmark I moved from folder A to folder B was back in folder A.

**Comment by Whissi** [comment]
In the past weeks I tried several older versions and I didn't find a single one working (oldest version I tried was v4.15). So for me I declare floccus completely broken on firefox. It's not usable anymore, especially because you can lose data.

**Comment by marcelklehr** [comment]
I'm sorry I've been absent from this issue.

> I believe that the thing with the button in navbar or not working bookmark dialog is caused by floccus getting stuck in syncing mode which will block bookmark API from firefox affecting firefox own navbar and other features like the bookmark dialog which all need to query bookmark data.

This is a false belief, AFAIK. Floccus doesn't block Firefox's bookmarks access.

> It's not usable anymore, especially because you can lose data.

I use it daily on Firefox and it works fine for me.

> So I cleared cache in floccus

This is what makes you lose data. There is a warning attached to that button I believe.

Can you tell me why this is not the same issue as https://github.com/floccusaddon/floccus/issues/1290 ?

**Comment by Whissi** [comment]
> This is a false belief, AFAIK. Floccus doesn't block Firefox's bookmarks access.

Well, this is at least how the error manifests for me. Disabling floccus add-in while firefox is "blocked" will "unblock" firefox immediately for me. I.e. on reload/new tab, blue bookmark star or bookmark menu works for me again without the need to restart the browser.

> This is what makes you lose data. There is a warning attached to that button I believe.

That's not how I read/understand

> Tick this box to reset the cache so that the next synchronization run **is guaranteed** **_not_** **to delete any data** and merely merges server and local bookmarks together

(from advanced options)


> Can you tell me why this is not the same issue as https://github.com/floccusaddon/floccus/issues/1290 ?

Hard to tell. Few weeks ago I found another issue discussing a locking issues where you posted some commands to check lock status. I remember that I tried this when my browser was stuck but there wasn't any lock. Also, my understanding of #1290 is that no other client will be able to sync when there is a (stale) lock. However, floccus in my Microsoft Edge and Google Chrome instances is working fine and last sync state is current (i.e. not what @jdelker experienced in https://github.com/floccusaddon/floccus/issues/1290#issuecomment-1348448541).

I would start by fixing the "Cancel" button (this is not meant to be instructive). I.e. when floccus is stuck in "syncing", "Cancel" button has no effect. I would expect at least to stop running action and reset floccus UI state. I would expect that this would remove any stale lock set by this client. At least I should be able to press cancel and start a new sync. At the moment, once floccus is stuck in syncing mode, there is no escape. I have to either clear cache or kill add-in.


I probably need to learn how to run this add-in from folder (not as XPI) so I can start debugging.

**Comment by 424344** [comment]
@Whissi  do you see any errors in the dev console of firefox? 
(to open it press CTRL+SHIFT+J. maybe you should select multiprocess option).

**Comment by Whissi** [comment]
Nothing in dev console when floccus is getting stuck, I only see the starting message: "Starting sync process for account ...". Also nothing when I press "Cancel" button to stop sync in progress. I wonder if this button works for anyone...

Don't know if I found something new: Since I lost bookmarks due to this issue, I stopped using floccus for a while. I now re-enabled floccus (4.18.1) on Firefox 109.0.1 and at least it looks like I don't need to reset cache to recover. Looks like it is enough to restart Firefox.

**Comment by Whissi** [comment]
Forget my previous comment. Because the problem hasn't happened that often in the past few days I thought that something could have changed in 4.18.x or Firefox 109 but it hasn't

Today I experienced the same issue like before: Console is showing "Starting sync process for account ..." and floccus became stuck. Firefox was showing bookmark icon for any URL I visited in navbar, bookmark menu didn't work...

While Firefox/floccus was stuck I started a Chromium-based browser where I also have floccus installed which is using same account and it had no problems to sync.

**Comment by silopolis** [comment]
Same here with an error 423 "Resource is already locked", possibly related to #1290

**Comment by Whissi** [comment]
FYI: I haven't experienced the problems since I changed storage of the volume where I store my Firefox profile (I used an RAID 10 on Adaptec SmartRAID 3154-8 before but for some unknown and unrelated reasons, this storage is sometimes very slow -- the whole system for example was hanging for ~2min after each cold boot. At this time my Windows user profile was stored on that volume. Since I moved my user profile to a NVMe disk I never experienced the problem again. This let me believe that there might be race condition you experience only on slow/overloaded storage system).
